### ISSUE TITLE
`dotnet restore` never completes

### ISSUE THREAD

**Issue by CollinAlpert** [dotnet/sdk#44373]
This one is a bit difficult to reproduce and to explain. Once in a while, when I execute `dotnet restore` on the command line, I get the "Determining projects to restore..." output, followed by "error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json" after a few minutes, but after that nothing happens and the command never completes. Restarting the computer always solves the issue and `dotnet restore` completes without any problems, however a restart is the only solution I have found so far.

As far as I can tell this happens at random times and I haven't been able to correlate to any certain event reliably. Here are a few of my observations, however please take my confirmation bias into account:
- Oftentimes when the issue occurred, I had been working on dotnet/roslyn-analyzers the day before.
- I first encountered the issue sometime after installing the preview version of .NET 7. 
- The issue always arises after being away from my computer for a few hours. It never worked and then stopped working while I was sitting at the computer, so I am thinking maybe something "Sleep"-related (in terms of the computer).
- Maybe this is Apple Silicon related? I have a Windows machine and have never had this issue there, even when having Preview versions installed and working on dotnet/roslyn-analyzers.
- There is no fixed time after the computer restart when the issue resurfaces. Sometimes everything will work for a few weeks, sometimes only for a couple of days.

I am just dumping this here in the hopes that someone a lot smarter than me will be able to make sense of it. Please let me know if you need more information or if you want me to try anything, I am desperate to get this resolved since it is very annoying.

<img width="720" alt="Screenshot 2023-10-09 at 18 36 48" src="https://github.com/dotnet/installer/assets/14217185/958c4240-22ac-4d97-a28d-a2575aa13af3">

`dotnet --info` output:
```
.NET SDK:
 Version:   8.0.100-rc.1.23463.5
 Commit:    e7f4de8816

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  13.6
 OS Platform: Darwin
 RID:         osx-arm64
 Base Path:   /Users/collin/.dotnet/sdk/8.0.100-rc.1.23463.5/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.0-rc.1.23419.4
  Architecture: arm64
  Commit:       92959931a3
  RID:          osx-arm64

.NET SDKs installed:
  8.0.100-rc.1.23463.5 [/Users/collin/.dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.13 [/Users/collin/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [/Users/collin/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-preview.6.23329.11 [/Users/collin/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-preview.7.23375.9 [/Users/collin/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.1.23421.29 [/Users/collin/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.13 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-preview.6.23329.7 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-preview.7.23375.6 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.1.23419.4 [/Users/collin/.dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/Users/collin/.dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

I am running macOS Ventura 13.6 with an M1 Pro chip.

**Comment by ctxnop** [comment]
I just had a similar issue, on Archlinux with version 10.0.100.

Any command involving networking (nuget, restore, workload, ...) just wait indefinitely without any message. I can kill the command but it still says nothing. I tried to leave it running for more than 10 minutes and when kill it I got some message about timeout on api.nuget.org.

I tried everything I found to increase verbosity but nothing gave me any hint. Could not use `dotnet-dump` because... well... the `dotnet tool install`  requires networking and so does hang too...

So I `strace`d it and found that it was hanging on trying to use IPv6 to connect remote servers, including api.nuget.org. The `connect` end up in a timeout after a while.

I don't know if the IPv6 connectivity is an issue on my side, I did not tried it, however I would say that every other apps seems to work (maybe those app just enforce IPv4, or correctly handle failure on IPv6), have no time and interest to investigate this more.

However, I solve the issue by editing the `/etc/gai.conf` file so that IPv4 takes precedence over IPv4.

**Comment by onionhammer** [comment]
Amazing this is still happening

**Comment by jozefizso** [comment]
We have the same issue with the dotnet 10.0.301 SDK. The `dotnet restore`, `dotnet build` or other commands never finish. Running them with diagnostics logging show nothing.


```
dotnet restore -v diag
  /Users/jozef/vbamc/vbamc.slnx                                                               Restore (100.3s)
```

dotnet info:

```
dotnet --info
.NET SDK:
 Version:           10.0.301
 Commit:            96856fd726
 Workload version:  10.0.300-manifests.b0c14421
 MSBuild version:   18.6.4+96856fd72

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  26.5
 OS Platform: Darwin
 RID:         osx-arm64
 Base Path:   /usr/local/share/dotnet/sdk/10.0.301/

.NET workloads installed:
There are no installed workloads to display.
Configured to use workload sets when installing new manifests.
No workload sets are installed. Run "dotnet workload restore" to install a workload set.

Host:
  Version:      10.0.9
  Architecture: arm64
  Commit:       901ca94124

.NET SDKs installed:
  8.0.416 [/usr/local/share/dotnet/sdk]
  9.0.308 [/usr/local/share/dotnet/sdk]
  10.0.301 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.22 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.11 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 10.0.9 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.22 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.11 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 10.0.9 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by onionhammer** [comment]
I've found if you kill all `aspire-managed` zombies it starts to work again

**Comment by baronfel** [comment]
It seems like the common thread is hanging on restore - @dotnet/nuget-team is there any way to diagnose mid-build what the Restore Task is currently working on/blocked doing? Are there known networking hangs/timeouts with IPv6 that should/could be added that might prevent these observed 'infinite' hangs?

**Comment by jozefizso** [comment]
> I've found if you kill all `aspire-managed` zombies it starts to work again

This sounds plausible. I restarted my Mac and it fixed the issue. I use Aspire on some other project and it could be stucked in the background.

**Comment by zivkan** [comment]
>  is there any way to diagnose mid-build what the Restore Task is currently working on/blocked doing?

I don't think I've ever had to do it in years I've been in the team. So, my best guess is to get a full memory dump of the process, so we can inspect it in the debugger. I've only tried once or twice to debug a memory dump captured on linux, and my memory is that Visual Studio couldn't open it, but maybe I'm wrong about that.

Visual Studio (not VSCode) has a very good "parallel stacks" view, and you can switch it from threads view to tasks view. Tasks view is what needs a full memory dump of heap memory, so a "minidump" will not have sufficient information. Also, I don't know if Parallel Stacks is a paid only feature, or if it's available in Visual Studio's Community edition. Once the memory dump is loaded in the Parallel Stacks task view, it will show which tasks are still pending, it will detect any deadlocks and highlight them with red symbols. I don't know how good the publicly available symbols Microsoft makes available are, but Microsoft employees using the internal symbol server will certainly be able to load the pdbs to see exactly what method each call stack is waiting on. It's a good tool for investigating what is being waited for at any moment in time.

I don't have experience debugging memory dumps outside of VS.

> Are there known networking hangs/timeouts with IPv6 that should/could be added that might prevent these observed 'infinite' hangs?

NuGet uses .NET's HttpClient in what I consider to be a very standard way. NuGet doesn't have any code that handles IP addresses directly, we just pass URLs straight to HttpClient and let it do DNS lookups, etc. So, I think this is a question better suited to the @dotnet/ncl team

**Comment by wfurt** [comment]
If you can get machine to state when it is failing, packet capture would be good way to check what is going on the network layer. I've seen security packages (like Defender) interfering with network traffic. There is deiagnocts logging but that may be hard to enable in the middle unless the restore/sdk has some provisioning for it. I'm wondering if it would make sense to add some debug/verbose option @zivkan and spin up EventListener that can dump diagnostic to a file or stderr. Also the never completes looks strange. HttpClient would try to connect ... but if that fails it would time out eventually - even if that can be long time. One can also hook up ConnectCallback to be able to monitor the connection and possibly write out diagnostic. There was also note about IPv6. On macOS (and other OSes) you can set DOTNET_SYSTEM_NET_DISABLEIPV6=1. IPv6 can still be  problematic in some environments.

**Comment by marcpopMSFT** [comment]
Old issue triage: This seems like a nuget restore issue and they need additional data to investigate further. Moving to the backlog for now.

**Comment by onionhammer** [comment]
This is a critical dev ex issue and should be routed to another team if you're not sure what's going on. Latent aspire-managed instances are locking things up in my case but to not even try is unforgivable

**Comment by zivkan** [comment]
I may be getting defensive here, but I think it's unfair to say we're not even trying.

There appear to be around 4 or 5 customers commenting that they're affected. My summary of this thread so far. Please let me know if I've missed something.

* Original poster said they don't have the problem on Windows, only on Apple Silicon. A second person saying they have the issue gave `dotnet --info` output, which shows they're on Mac
* Another person said they're having the issue on Arch Linux, but stopped having the problem when they reconfigured their system to prefer IPV4 over IPV6.
* Another person said they can't install `dotnet-dump` because `dotnet tool` needs networking to download/install it.
* You mentioned stopping `aspire-managed` processes makes it work again. This comment got 1 👍 reaction, but the person who reacted immediately commented saying they didn't try it, they rebooted their computer instead.
* The original poster provided a dump.txt, which contains a stack trace of every thread of every process on their machine. But since NuGet's networking is async, it's not on any threads, so it doesn't give me any clues I can find to investigate.
* A few comments from Microsoft employees pinging each other and saying things like memory dumps and network traces could help investigate.

I don't own any Apple Silicon, so I don't have a device to run tests on to see if I can reproduce the issue. Our work provided devices all run Windows, and the original poster said they also don't have the problem on Windows, only on Mac. One other person mentioned Linux, but otherwise I don't have additional information about what operating systems this issue happens on. I run Arch Linux on my personal devices, and I haven't had any problems with restore or IPV6.

While I get notifications on this issue, I don't know how many other people in the NuGet team do, so creating a new issue in NuGet/Home would make sure that the person on the rotating support roster should see it. But without more information on how to reproduce the issue or a full memory dump (not just stack traces, we need to look at incomplete async Tasks), I just don't know how to guess what might be wrong.

When NuGet makes an HTTP request with HttpClient, [it uses a timeout via a cancellation token](https://github.com/NuGet/NuGet.Client/blob/5b8014660bfd5051a0367d5a2e826918c66e262d/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs#L152-L163). If the HTTP headers are returned, then [Nuget also uses a timeout while reading the response body](https://github.com/NuGet/NuGet.Client/blob/5b8014660bfd5051a0367d5a2e826918c66e262d/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs#L173).  If either of these timeouts occur, NuGet will retry, default is up to 5 times. Each retry will be output at restore's normal and detailed verbosity, but [msbuild's terminal logger will hide all messages not on their allow-list](https://github.com/dotnet/sdk/issues/43293), so you'll need to disable the terminal logger to see the log messages. With default timeouts being 100 seconds, and 5 retries by default, plus a short cooldown between retries, it'll take a bit over 10 minutes for a single HTTP request to completely fail.  If the failure happens when trying to get the package source's service index (index.json), then restore should fail pretty quickly after the last retry fails. But if it passes that point, then other URLs might also be tried, and they'll also have 10+ minutes to fully fail. But this all assumes the problem is timeouts. The original post clearly shows errors saying the index.json couldn't be fetched. We've improved the error in the last year or two to show the inner exception to explain why the service index failed to download.

So, the network trace will help to figure out if progress is being made very slowly, preventing NuGet's timeout logic from working. A full memory dump will help us validate that there are downloads in progress, or if something else is being waited on, but I don't think a memory dump will let us calculate how long the wait has been happening, for example to check if a timeout trigger failed to fire. But I think we might be able to figure out if a cancellation token is requesting cancellation and a waiting task hasn't yet entered a failed state.

From my point of view, I don't have instructions on how to reproduce the issue, and I've never experienced such an issue. NuGet's code already had timeouts built into the networking calls.  I'm not sure what more I can do until more information is provided.

**Comment by zivkan** [comment]
> The original poster provided a dump.txt, which contains a stack trace of every thread of every process on their machine. But since NuGet's networking is async, it's not on any threads, so it doesn't give me any clues I can find to investigate.

Just after clicking send, I had a thought: MSBuild tasks are not async, so I would expect at least one thread to be running NuGet.Build.Tasks.RestoreTask.Execute, being called from some Microsoft.Build.* code. But neither MSBuild nor NuGet appear in the dump.txt or sample.txt files. Therefore, from the evidence provided so far, it's pointing to something that's neither NuGet nor MSBuild.

Hence we need more information to investigate further.

**Comment by ctxnop** [comment]
> - Another person said they're having the issue on Arch Linux, but stopped having the problem when they reconfigured their system to prefer IPV4 over IPV6.
> - Another person said they can't install dotnet-dump because dotnet tool needs networking to download/install it.

The two guys you mention are... the same one. Me. Inside the same comment.

And you understood the "I workaround it by setting preference on IPv4" as "it was a bad configured archlinux, not a bug". Workaround is not a fix, and I want my system to prefer IPv6 over IPv4, I do not like to have to make such a system-wide change to make such a user tool to work. I consider the issue still here and I'm waiting for a fix so that I can restore the correct configuration of my system.

I get that you didn't read carefully through the comments (me neither, but I'm not the one who is trying to fix the issue), or asked copilot to summarize it, and you wait for new comments for more info. Instead of waiting it to happen, you can ask people to provide the infos you want, you can try to build a setup to reproduce, etc...

You say you don't have an Apple but you do have an Archlinux, however you don't reproduce the issue on it. Good, as it is fixed by a system wide configuration, you might expect that the issue comes only after you installed this or that package, or when you have this or that configuration changed. Hence you should start with a container, or a VM, and discuss with people (as the one who reported the issue on archlinux it would be me) to get infos to approximate his setup. Like, did you even check the file I mentioned (`/etc/gai.conf`) to see if it exists on you setup and if it configured to prefer IPv6 over IPv4? Because if you already have the same workaround as I have, then no wonder why it works for you.

**Comment by zivkan** [comment]
I see that previous comments from Microsoft employees answered questions on "what information would be useful", rather than explicitly asking for specific information, which is a fair criticism. Anyone experiencing the issue, please try to get the following information:

1. Network trace
   * Ideally in a file format that's widely supported, such as `.pcap` or whatever Wireshark's modern file format is.
2. A full memory dump
   * Windows (or maybe just .NET on Windows) has the concept of a minidump, which just has threads and callstacks. But it misses async tasks, so a full memory dump is needed.
   * On Windows that can be done with Sysinternals procdump, or Task Manager. The dotnet-dump tool 
   * If `dotnet tool install --global dotnet-dump` doesn't work because of networking issues, you can go to https://www.nuget.org/packages/dotnet-dump in your web browser, click the download package link, then use the `--source` command, providing the directory the nupkg is saved in.

Keep in mind any attachments here on GitHub can be downloaded by anyone, so if you have a Microsoft account or a work Entra ID (previously called Azure Active Directory) account, you can report the issue with attachments at https://developercommunity.visualstudio.com/VisualStudio/report?webReport=true, and only Microsoft employees will be able to download the attachments.


I probably shouldn't, but addressing the "you could have" suggestions, like most people I have more to do than time to do it. Prioritization decisions have to be made, and that includes taking into account if an issue is actionable, how many customers will benefit, and the severity/impact of the issue or new feature. Commands that never complete are high severity, but if it was a widespread issue, I'd expect more upvotes, more comments, and more duplicate issues. I know how much it sucks when something doesn't work for me, and few other people have the same problem, so I can't convince the developers to investigate and fix the issue. But I still have more actionable issues to investigate or more impactful features to implement.

FWIW, my /etc/gai.conf is just comments. AI chatbots tell me this means it's using glibc's default of preferring IPV6 over IPV4.

**Comment by onionhammer** [comment]
If lots of people are getting this same behavior, but for different reasons or undiagnosed reasons it points to the code being brittle. It's also an ephemeral issue that's difficult to reproduce, but people are rebooting their machines in order to get dotnet restore to work - This should be concerning

I'm not blaming you for not having apple silicon, but it's a supported platform, so maybe there's another team member who does or might know what's going on

**Comment by ocdi** [comment]
I can 100% confirm force killing stuck `aspire-managed` processes fixed this without rebooting. They were in state `Tl`. How they ended up like this is beyond me and unfortunately I didn't capture anything further. These remaining 4 required SIGKILL and I had 5 others that exited on SIGTERM. I wonder if there is some kind of shared lock that could cause this? I'm using Linux and .NET 10 currently.

```
PID    PPID STAT     ELAPSED CMD
2694232    3895 Tl    3-00:09:58 /home/t/.aspire/versions/13.4.5_73114e86c64aeb9f3f3c7da8e37df1ae4281b27e-91a12bb7860c4e44/managed/aspire-managed nuget search --query Aspire.Cli --take 1000 --format json --prerelease --working-dir /home/t/code/project1 --verbose
3166953    3895 Tl      15:12:41 /home/t/.aspire/versions/13.4.5_73114e86c64aeb9f3f3c7da8e37df1ae4281b27e-91a12bb7860c4e44/managed/aspire-managed nuget search --query Aspire.ProjectTemplates --take 1000 --format json --working-dir /home/t/code/project2 --verbose
3258703    3895 Tl      14:55:54 /home/t/.aspire/versions/13.4.5_73114e86c64aeb9f3f3c7da8e37df1ae4281b27e-91a12bb7860c4e44/managed/aspire-managed nuget search --query Aspire.Cli --take 1000 --format json --prerelease --working-dir /home/t/code/project2 --verbose
3338990    3895 Tl      14:39:20 /home/t/.aspire/versions/13.4.5_73114e86c64aeb9f3f3c7da8e37df1ae4281b27e-91a12bb7860c4e44/managed/aspire-managed nuget search --query Aspire.Cli --take 1000 --format json --prerelease --working-dir /home/t/code/project2 --verbose
```

**Comment by ocdi** [comment]
I believe this bug is actually in the aspire CLI, the linked issue above describes my findings with the assistance of AI. The ultimate solution (as already stated above) is to forcibly kill these stuck processes to release the lock.

**Comment by zivkan** [comment]
The aspire issue appears to have been fixed a few days ago, so hopefully a new release will solve it soon:
* https://github.com/microsoft/aspire/issues/18948

However, this issue was opened in 2023, and Aspire appears to have been first released in early 2024, so it feels unlikely that everyone is experiencing the same problem.
